### ISSUE TITLE
Kubernetes end to end tests should run on Macs.

### ISSUE THREAD

**Issue by StephenWakely** [timberio/vector#7545]
Currently the tests assume that you are using Linux. This allows them to build Vector locally and deploy a container to test to Docker Hub. 

On a Mac you have to deploy the Vector container to Docker Hub yourself and then specify `SKIP_CONTAINER_IMAGE_PUBLISHING=true CONTAINER_IMAGE=<image>` when running the tests. We should build it into the tests to use Docker to build a Vector image for Macs. There is also a possibility we could use cross compiling, but I couldn't get it to work - there were issues with some of the third party libraries we use.

**Comment by spencergilbert** [comment]
I think this can be closed, or maybe we need to add a note that it requires GNU make of ~4.3, waiting to confirm with @StephenWakely and @prognant

**Comment by jszwedko** [comment]
I wonder if we could add a check in the `Makefile` itself to verify compatibility and error early if the version of `make` is too old.

**Comment by prognant** [comment]
It seems indeed that `make` shipped with big sur (3.81) does not work (some kind of infinite loop target, https://gist.github.com/prognant/7728767f89d95004673fa81c114cc564) while the one from brew (4.3) seems to work (at least it go further) if possible I would say we should try to update the Makefile so it works with make versions used out there (I have no idea if that's possible).

**Comment by spencergilbert** [comment]
Is it purely a version issue or a BSD vs GNU problem?

**Comment by StephenWakely** [comment]
I received this error after several hourse of building: 

```
error: could not compile `k8s-openapi`

Caused by:
  process didn't exit successfully: `rustc --crate-name k8s_openapi --edition=2018 /cargo/registry/src/github.com-1ecc6299db9ec823/k8s-openapi-0.11.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C linker-plugin-lto -C codegen-units=1 --cfg 'feature="api"' --cfg 'feature="default"' --cfg 'feature="http"' --cfg 'feature="percent-encoding"' --cfg 'feature="url"' --cfg 'feature="v1_16"' -C metadata=5611f29a59d6d2dc -C extra-filename=-5611f29a59d6d2dc --out-dir /target/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/target/x86_64-unknown-linux-gnu/release/deps -L dependency=/target/release/deps --extern base64=/target/x86_64-unknown-linux-gnu/release/deps/libbase64-6061d76424c3816a.rmeta --extern bytes=/target/x86_64-unknown-linux-gnu/release/deps/libbytes-cd5092bb6fd0e2f2.rmeta --extern chrono=/target/x86_64-unknown-linux-gnu/release/deps/libchrono-201f9b240215be4d.rmeta --extern http=/target/x86_64-unknown-linux-gnu/release/deps/libhttp-fb1421e8e1952222.rmeta --extern percent_encoding=/target/x86_64-unknown-linux-gnu/release/deps/libpercent_encoding-45dc6f248b908bcb.rmeta --extern serde=/target/x86_64-unknown-linux-gnu/release/deps/libserde-76586474eebda37b.rmeta --extern serde_value=/target/x86_64-unknown-linux-gnu/release/deps/libserde_value-9b0283a13b017c04.rmeta --extern serde_json=/target/x86_64-unknown-linux-gnu/release/deps/libserde_json-7d5d8d6dc64a6c50.rmeta --extern url=/target/x86_64-unknown-linux-gnu/release/deps/liburl-0101410dca9c8a4d.rmeta --cap-lints allow -C link-args=-rdynamic` (signal: 9, SIGKILL: kill)
warning: build failed, waiting for other jobs to finish...
error: build failed
make[1]: *** [Makefile:238: target/x86_64-unknown-linux-gnu/release/vector] Error 101
make[1]: Leaving directory '/Users/stephen.wakely/src/vector'
make: *** [Makefile:495: test-e2e-kubernetes] Error 2
```

Am trying to build again, but it seems exceedingly slow (at least several hours). Should that be the case?

**Comment by jszwedko** [comment]
> if possible I would say we should try to update the Makefile so it works with make versions used out there (I have no idea if that's possible).

We do have `make environment` which will drop you into a docker container with, hopefully, the correct toolchains installed. This is a problem I've dealt with at many companies and have never come across a great solution 😄 A preconfigured development VM is the best I've seen but has a lot of overhead.

We had a brief foray into nix which, as I understand it, could also help with this issue, but ended up backing it out as we couldn't get it working completely.

>   process didn't exit successfully: `rustc --crate-name k8s_openapi --edition=2018 /cargo/registry/src/github.com-1ecc6299db9ec823/k8s-openapi-0.11.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C linker-plugin-lto -C codegen-units=1 --cfg 'feature="api"' --cfg 'feature="default"' --cfg 'feature="http"' --cfg 'feature="percent-encoding"' --cfg 'feature="url"' --cfg 'feature="v1_16"' -C metadata=5611f29a59d6d2dc -C extra-filename=-5611f29a59d6d2dc --out-dir /target/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/target/x86_64-unknown-linux-gnu/release/deps -L dependency=/target/release/deps --extern base64=/target/x86_64-unknown-linux-gnu/release/deps/libbase64-6061d76424c3816a.rmeta --extern bytes=/target/x86_64-unknown-linux-gnu/release/deps/libbytes-cd5092bb6fd0e2f2.rmeta --extern chrono=/target/x86_64-unknown-linux-gnu/release/deps/libchrono-201f9b240215be4d.rmeta --extern http=/target/x86_64-unknown-linux-gnu/release/deps/libhttp-fb1421e8e1952222.rmeta --extern percent_encoding=/target/x86_64-unknown-linux-gnu/release/deps/libpercent_encoding-45dc6f248b908bcb.rmeta --extern serde=/target/x86_64-unknown-linux-gnu/release/deps/libserde-76586474eebda37b.rmeta --extern serde_value=/target/x86_64-unknown-linux-gnu/release/deps/libserde_value-9b0283a13b017c04.rmeta --extern serde_json=/target/x86_64-unknown-linux-gnu/release/deps/libserde_json-7d5d8d6dc64a6c50.rmeta --extern url=/target/x86_64-unknown-linux-gnu/release/deps/liburl-0101410dca9c8a4d.rmeta --cap-lints allow -C link-args=-rdynamic` (signal: 9, SIGKILL: kill)

This looks like an OOM kill. I believe cross building on OSX for Linux will cause it to run the build in docker on the docker for mac VM. You could try giving it more resources. The default config I see on my machine is 4 CPUs and 2 GB of memory which is definitely not much to build vector 😄 I'd try bumping it to at least 16 GB.

**Comment by spencergilbert** [comment]
FWIW I've given docker 4cpu and 12gb and seems to work, albeit slowly

**Comment by jszwedko** [comment]
I also still maintain that we don't need to do release builds for the end-to-end correctness tests which should cut down on cycle time a lot.

**Comment by spencergilbert** [comment]
With regards to this issue and #7173 - do we want to just document that we need a GNU Make ~v4.3 or do we want to spend some time to either change the makefile to support older/non-GNU versions OR check make and error out with a useful message?

**Comment by jszwedko** [comment]
My suggestion would be to make it compatible with the default make on OSX if it is a low effort change (< a couple of hours) otherwise I'd add the check to output a better error message.

**Comment by spencergilbert** [comment]
There's definitely something fiddly going on here as I've successfully run: `SCOPE=simple /usr/bin/make test-e2e-kubernetes`

```bash
❯ /usr/bin/make --version
GNU Make 3.81
```

**Comment by prognant** [comment]
I did a bunch of tests on my mac, few conclusion:
* Minikube non-default profile name was causing trouble (fix here #7853)
* The `make` coming with macosx is not causing trouble anymore (#7844)
* As pointed by @jszwedko, steps ending up by a task getting killed were (at least on my mac) terminated because of OOM. Having 8G of RAM allocated to the docker hyperkit VM is enough to successfully complete `make test-e2e-kubernetes` (the container used by the build process are topping around ~5.7Gb of RAM)

I don't think there is much to add, maybe we could add a minimal resources requirements in `CONTRIBUTING.md` ?

**Comment by spencergilbert** [comment]
> * As pointed by @jszwedko, steps ending up by a task getting killed were (at least on my mac) terminated because of OOM. Having 8G of RAM allocated to the docker hyperkit VM is enough to successfully complete `make test-e2e-kubernetes` (containers used by the build process are topping around ~5.7Gb of RAM)
> 
> I don't think there is much to add, maybe we could add a minimal resources requirements in `CONTRIBUTING.md` ?

Sounds good to me, I've been using 2cpu and 8gb without issue so far
